lib/collection/src/update_handler.rs

```diff
@@ -285,18 +335,27 @@ impl UpdateHandler {
         optimization_handles: Arc<TokioMutex<Vec<StoppableTaskHandle<bool>>>>,
         max_handles: usize,
     ) {
-        while let Some(signal) = receiver.recv().await {
-            match signal {
-                OptimizerSignal::Nop | OptimizerSignal::Operation(_) => {
+        loop {
+            let receiver = timeout(OPTIMIZER_CLEANUP_INTERVAL, receiver.recv());
+            let result = receiver.await;
+
+            // Always clean up on any signal
+            Self::cleanup_optimization_handles(optimization_handles.clone()).await;
```

Comment:
Pointer: the worker now waits for any signal, but also continues the loop every 5 seconds.

This way we always clean up finished optimization tasks at least every 5 seconds, as that is done right after. This is needed to catch and propagate panics in time.

If we actually received a signal, it is handled below.

Reply:
my concern here is that current behaviour is so, that if optimisation caused oomkill, it won't be automatically started after the restart, potentially preventing the crash loop. This might be sub-optimal in some cases, as the downside is that optimisation might appear not starting without a reason

Reply:
It is not entirely clear to me what you mean. Do you mean that this may catch an OOM panic without restarting?

Doesn't an OOM event termiate the process immediately without unwinding?

Reply:
What I mean, that starting optimization right after restart may cause crashloop

Reply:
Thanks for elaborating. I don't think it immediately starts optimization.

The signal receiver with timeout does not start optimization when it times out. Instead it only cleans up finished optimization handles (which would be none) and then waits for a signal [again](https://github.com/qdrant/qdrant/blob/d8d2306a9b176e78c8e48d720d2874938db1243d/lib/collection/src/update_handler.rs#L376-L377).

In other words, an optimization is [only](https://github.com/qdrant/qdrant/blob/d8d2306a9b176e78c8e48d720d2874938db1243d/lib/collection/src/update_handler.rs#L378-L401) started if a `OptimizerSignal::Nop` or `OptimizerSignal::Operation(_)` is received through the channel.

Or am I missing something important here?

Reply:
I can see it now, thanks! I think my concern is resolved then